### DISCUSSION TITLE
Does UDP channel support retransmissions?

### DISCUSSION THREAD

**Discussion by bizzbyster** [vpnhood/VpnHood#95] [Q&A] [answered]
According to your wiki (https://github.com/vpnhood/VpnHood/wiki/How-does-VpnHood-work), you are terminating TCP connections at the local client and then establishing a new TCP connection between the VpnHoodServer and the content server. When the Vpn uses TCP between VpnHoodClient and VpnHoodServer, this makes sense to me. But I'm struggling to understand how the UDP channel works for the case where packets are lost between client and server. Have you added some sort of retransmission mechanism so that the UDP channel is reliable?

**Comment by trudyhood** [comment, marked as answer]
Firstly, keep in mind that UDP is not crucial for some people who suffer severe filtering except for DNS resolving and a few other services of-course many services like youtube try to use that, but when it is not available, they fall back to TCP mode.

By the way, In TCP mode, VpnHood establishes a few HTTPS connections and disables the TCP Nagle algorithm (NO DELAY). We named it TcpDatagramChannel!. Then VpnHood encapsulates every UDP and sends them as TCP data through those channels like ordinary data, VpnHood does some optimization like batching packets and find a free TcpDatagramChannel to send the UDP data, because we don't need the reliability of TCP connection here, it sends UDP packets through multiple TcpDatagramChannels and does not wait for TCP ACK. I mean when you send some data over TCP, the connection halt until receives ACK for reliability, meanwhile VpnHood sends UDP packets through other already free established TcpDatagramChannels to boost the performance.

**Have you added some sort of retransmission mechanism so that the UDP channel is reliable?**
The fun part is that we don't need that. Any app that uses UDP expects that UDP is not reliable, so like a real environment, if there is any packet loss, the apps retry sending them by itself. Even, VpnHood sends the UDP packet more reliable than usual because it sends UDP through TCP! However, it decreases the performance because we should not make UDP reliable.

**Comment by bizzbyster** [comment]
Thanks @trudyhood. That all makes sense. I'm still unclear on something - when I choose "UDP Mode" in the UI, it still uses TCP in the tunnel between client and server for the application data that is sent using TCP? Sorry if I'm not quite getting it.

**Reply by trudyhood** [reply]
Dear @bizzbyster 
Feel free to ask any confusion you have.

In (no UDP) mode: VpnHood uses TCP proxy for TCP connections and uses TCP connection (TcpDatagram) for UDP packets.
In (use UDP) mode: VpnHood uses TCP proxy for TCP connections but uses UDP for UDP packets. In this case, VpnHood encapsulates UDP packets to another UDP packets and sent them to the VpnHood server, the same as ordinary VPNs, but this happens just for UDP packets. Still, It establishes TCP connections through TCP proxies. In other words, UDP mode does nothing with TCP connections.
